suite/tests/client-interface/drbbdup-emul-reg-clobber-test.dll.c

```diff
@@ -0,0 +1,311 @@
+#include "client_tools.h"
+#include "drmgr.h"
+#include "drbbdup.h"
+#include "drutil.h"
+#include "drreg.h"
+#include <string.h>
+
+typedef struct _per_block_t {
+    bool saw_first;
+    bool saw_last;
+} per_block_t;
+
+/* Assume single threaded. */
+static uintptr_t encode_val = 1;
+
+#ifdef X86
+static bool saw_movs;
+#endif
+static bool has_rest_of_block_emul;
+
+static dr_emit_flags_t
+app2app_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+              bool translating)
+{
+#if VERBOSE
+    /* Useful for debugging. */
+    instrlist_disassemble(drcontext, tag, bb, STDERR);
+#endif
+
+    has_rest_of_block_emul = false;
+
+    /* Test drutil rep string expansion interacting with drbbdup. */
+    bool expanded;
+    if (!drutil_expand_rep_string_ex(drcontext, bb, &expanded, NULL))
+        DR_ASSERT(false);
+    if (expanded) {
+        /* We can't overlap ours with drutil's so bow out. */
+        has_rest_of_block_emul = true;
+        return DR_EMIT_DEFAULT;
+    }
+    /* Test handling of DR_EMULATE_REST_OF_BLOCK by inserting it in the middle
+     * of a multi-instr block and ensuring drbbdup doesn't let it leak through into
+     * the start of subsequent cloned blocks.
+     * We want to leave some blocks unchanged though to test drbbdup's own
+     * emulation labels.
+     *
+     * XXX i#5390: We should also test the last instr being emulated: not sure drbbdup
+     * will do the right thing there if it's a "special" instr.
+     */
+    static int bb_count;
+    if (++bb_count % 2 == 0)
+        return DR_EMIT_DEFAULT;
+    instr_t *mid = instrlist_last_app(bb);
+    if (mid == NULL || mid == instrlist_first_app(bb))
+        return DR_EMIT_DEFAULT;
+    mid = instr_get_prev_app(mid);
+    if (mid == NULL || mid == instrlist_first_app(bb))
+        return DR_EMIT_DEFAULT;
+    emulated_instr_t emulated_instr;
+    emulated_instr.size = sizeof(emulated_instr);
+    emulated_instr.pc = instr_get_app_pc(mid);
+    emulated_instr.instr = instr_clone(drcontext, mid);
+    emulated_instr.flags = DR_EMULATE_REST_OF_BLOCK;
+    drmgr_insert_emulation_start(drcontext, bb, mid, &emulated_instr);
+    /* XXX i#5400: We'd like to pass a user_data to the instrument event but drbbdup
+     * doesn't support that; instead we rely on being single-threaded.
+     */
+    has_rest_of_block_emul = true;
+
+    return DR_EMIT_DEFAULT;
+}
+
+static uintptr_t
+set_up_bb_dups(void *drbbdup_ctx, void *drcontext, void *tag, instrlist_t *bb,
+               bool *enable_dups, bool *enable_dynamic_handling, void *user_data)
+{
+    drbbdup_status_t res;
+
+    CHECK(enable_dups != NULL, "should not be NULL");
+    CHECK(enable_dynamic_handling != NULL, "should not be NULL");
+
+    res = drbbdup_register_case_encoding(drbbdup_ctx, 1);
+    CHECK(res == DRBBDUP_SUCCESS, "failed to register case 1");
+    res = drbbdup_register_case_encoding(drbbdup_ctx, 2);
+    CHECK(res == DRBBDUP_SUCCESS, "failed to register case 2");
```

Comment:
I thought the plan was to add a 3rd case to trigger flags reservation in drbbdup.

Reply:
I believe that there are three cases here: 0 (default), 1, and 2.

Also, the problem occurs when the instrumentation modifies flags in a non-default drbbdup case. The flag reservation by drbbdup does not cause the problem.